// ngm-cli/commands/build.command.js
// todo: add load from config file, TBD
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Listr = require('listr');
const cpy = require('cpy');
const del = require('del');
const npm_submodules_1 = require("npm-submodules");
const tasks_1 = require("../tasks");
function buildCommand({ project, verbose, clean, local, main, watch, skipBundles }) {
    // 1. clean dist folders
    // 2.1 merge pkg json
    // todo: 2.2 validate pkg (main, module, types fields)
    // 2.3 write pkg
    // 3. compile ts
    return npm_submodules_1.findSubmodules(project, { local })
        .then(opts => new Listr([
        {
            title: 'Clean dist folders',
            task: () => new Listr(opts.map(opt => ({
                title: `Cleaning ${opt.dist}`,
                task: () => del(opt.dist)
            }))),
            skip: () => !clean
        },
        {
            title: 'Copy md files and license',
            task: () => Promise.all(opts.map(opt => cpy(['*.md', 'LICENSE'], opt.dist)
                .then(() => cpy([path.join(opt.src, '*.md'),
                path.join(opt.src, 'LICENSE')], opt.dist))))
        },
        {
            title: "Build package.json files",
            task: () => npm_submodules_1.buildPkgs(opts, { local })
        },
        {
            title: 'Build projects',
            task: () => new Listr(opts.map(opt => ({
                title: `Building ${opt.pkg.name} (${opt.src})`,
                task: () => tasks_1.build(opt.project)
            })))
        },
        {
            title: 'Bundling umd version',
            task: () => new Listr(opts.map(opt => ({
                title: `Bundling ${opt.pkg.name}`,
                task: () => tasks_1.bundleUmd({
                    main,
                    src: opt.src,
                    dist: opt.dist,
                    name: opt.pkg.name,
                    tsconfig: opt.tsconfig.path,
                    minify: false
                })
            }))),
            skip: () => watch && skipBundles
        },
        {
            title: 'Bundling minified umd version',
            task: () => new Listr(opts.map(opt => ({
                title: `Bundling ${opt.pkg.name}`,
                task: () => tasks_1.bundleUmd({
                    main,
                    src: opt.src,
                    dist: opt.dist,
                    name: opt.pkg.name,
                    tsconfig: opt.tsconfig.path,
                    minify: true
                })
            }))),
            skip: () => watch || skipBundles
        },
    ], { renderer: verbose ? 'verbose' : 'default' }));
}
exports.buildCommand = buildCommand;
function buildTsRun(cli) {
    const { project, watch, verbose, clean, local, skipBundles } = cli.flags;
    let main = cli.flags.main || 'index.ts';
    return buildCommand({ project, verbose, clean, local, main, watch, skipBundles })
        .then(tasks => npm_submodules_1.tasksWatch({ project, tasks, watch }));
}
exports.buildTsRun = buildTsRun;
